### ISSUE TITLE
Include checksum in versionInfo

### ISSUE THREAD

**Issue by sabbyanandan** [spring-cloud/spring-cloud-dataflow#1779]
As a developer, I'd like to add SHA-1 checksum info for core, dashboard, and shell returned in the `versionInfo`.

Note: This is a pre-requisite for spring-cloud/spring-cloud-dataflow#1758.

**Acceptance:**
- Investigate how to generate and extract JFrog artifactory's SHA-1 hash for all the SCDF artifacts including Shell
- Embed checksum info for core, dashboard, and shell 
- Amend REST docs to include the new attributes

**Sample:**
```
{
  "versionInfo": {
    "implementation": {
      "name": "spring-cloud-dataflow-server-cloudfoundry",
      "version": "1.3.0.BUILD-SNAPSHOT",
      "checksum-sha1": "",
      "checksum-sha256": "",
      "url": "http://....."
    },
    "core": {
      "name": "Spring Cloud Data Flow Core",
      "version": "1.3.0.BUILD-SNAPSHOT",
      "checksum-sha1": "",
      "checksum-sha256": "",
      "url": "http://....."
    },
    "dashboard": {
      "name": "Spring Cloud Dataflow UI",
      "version": "1.3.0.M3",
      "checksum-sha1": "",
      "checksum-sha256": "",
      "url": "http://....."
    },
    "shell": {
      "name": "Spring Cloud Dataflow Shell",
      "version": "1.3.0.M3",
      "checksum-sha1": "",
      "checksum-sha256": "",
      "url": "http://....."
    }
  }
}
```

**Comment by cppwfs** [comment]
A couple of questions come to mind.   
1) If the users are pulling from maven central, all they will have access to is md5 and sha1.  So where will the source of truth be for SHA256.  repo.spring.io doesn't really serve this purpose once SCDF is released to maven central.  
2) Once we determine the source of truth, there will have to be some configuration to where the about can search for the SHA so that it can retrieve it.

Thoughts?

**Comment by glyn** [comment]
Please note that SHA-256 was specified in https://github.com/spring-cloud/spring-cloud-dataflow/issues/1758 because MD5 and SHA-1 are no longer regarded as secure (collisions can be generated).

**Comment by sabbyanandan** [comment]
As an update, we will stick to SHA-1 (for now) until JFrog Artifactory adds support for SHA-256.

**Comment by glyn** [comment]
@sabbyanandan Please could you _document_ both SHA-1 and, optional, SHA-256 checksum fields so that I can ship the plugin code to accept SHA-256 automatically as soon as the dataflow server supports it and fall back to SHA-1 when SHA-256 is not available?

**Comment by cppwfs** [comment]
@glyn I'm adding the following configuration fields
* spring_cloud_dataflow_version-info_core_url - set the link to the data flow core artifact
* spring_cloud_dataflow_version-info_dashboard_url - set the link to the data flow dashboard artifact
* spring_cloud_dataflow_version-info_implementation_url - set the link to the data flow server artifact
* spring_cloud_dataflow_version-info_shell_url - set the link to the data flow shell artifact
* spring_cloud_dataflow_version-info_type - the checksum type to be suffixed to the artifact links enumerated in the list above.   

An example configuration as show  below: 
`spring_cloud_dataflow_version-info_shell_url : "{repository}/org/springframework/cloud/spring-cloud-dataflow-shell/{version}/spring-cloud-dataflow-shell-{version}.jar"`
`spring_cloud_dataflow_version-info_type: "sha1"`
would produce the following link if the version is BUILD-SNAPSHOT in the about json.
```
    "shell": {
      "name": "Spring Cloud Data Flow Shell",
      "version": "1.3.0.BUILD-SNAPSHOT",
      "checksum": "bfcd1b73089b8f07b2a201036865b25a4624ab19",
      "url": "https:\/\/repo.spring.io\/libs-snapshot\/org\/springframework\/cloud\/spring-cloud-dataflow-shell\/1.3.0.BUILD-SNAPSHOT\/spring-cloud-dataflow-shell-1.3.0.BUILD-SNAPSHOT.jar"
    }
```

**Comment by glyn** [comment]
@cppwfs I see, thanks. What if you need to set both a SHA-1 and a SHA-256?

The plugin code needs to know the type of checksum it has received. JSON like this would help:
```
"shell": {
      "name": "Spring Cloud Data Flow Shell",
      "version": "1.3.0.BUILD-SNAPSHOT",
      "checksum-sha1": "bfcd1b73089b8f07b2a201036865b25a4624ab19",
      "url": "https:\/\/repo.spring.io\/libs-snapshot\/org\/springframework\/cloud\/spring-cloud-dataflow-shell\/1.3.0.BUILD-SNAPSHOT\/spring-cloud-dataflow-shell-1.3.0.BUILD-SNAPSHOT.jar"
 }
```

Thoughts?

**Comment by cppwfs** [comment]
@glyn to clarify do you want each artifact(artifact meaning: shell, core, dashboard, etc) to select its own checksum ?  
```
"shell": {
      "name": "Spring Cloud Data Flow Shell",
      "version": "1.3.0.BUILD-SNAPSHOT",
      "checksum-sha1": "bfcd1b73089b8f07b2a201036865b25a4624ab19",
      "url": "https:\/\/repo.spring.io\/libs-snapshot\/org\/springframework\/cloud\/spring-cloud-dataflow-shell\/1.3.0.BUILD-SNAPSHOT\/spring-cloud-dataflow-shell-1.3.0.BUILD-SNAPSHOT.jar"
 }
"dashboard": {
      "name": "Spring Cloud Data Flow Dashboard",
      "version": "1.3.0.BUILD-SNAPSHOT",
      "checksum-sha256": "magicsha256",
      "url": "https:\/\/repo.spring.io\/libs-snapshot\/org\/springframework\/cloud\/spring-cloud-dataflow-shell\/1.3.0.BUILD-SNAPSHOT\/spring-cloud-dataflow-dashboard-1.3.0.BUILD-SNAPSHOT.jar"
 }
```

Or do you want each artifact to show multiple checksums i.e. ?
"shell": {
      "name": "Spring Cloud Data Flow Shell",
      "version": "1.3.0.BUILD-SNAPSHOT",
      "checksum-sha1": "bfcd1b73089b8f07b2a201036865b25a4624ab19",
      "checksum-sha256": "abcd...",
      "url": "https:\/\/repo.spring.io\/libs-snapshot\/org\/springframework\/cloud\/spring-cloud-dataflow-shell\/1.3.0.BUILD-SNAPSHOT\/spring-cloud-dataflow-shell-1.3.0.BUILD-SNAPSHOT.jar"
 }

**Comment by sabbyanandan** [comment]
Glyn only requires the specs for Shell. We are adding it to others for consistency. And, yes, we can add `checksum-sha256` and `checksum-sha1` both. It might be odd to see `checksum-sha256` empty (for a while), but it doesn't add any overhead.

**Comment by cppwfs** [comment]
I'll add a checksum-sha1 and checksum-sha256.   Just wanted clarification.

**Comment by glyn** [comment]
@sabbyanandan Thanks for the clarification. Would it be legitimate to omit `checksum-sha256` altogether until it is supported? That seems cleaner than including it with an invalid, albeit empty, value.

**Comment by glyn** [comment]
@cppwfs At the risk of confusing matters, my question:

> What if you need to set both a SHA-1 and a SHA-256?

was talking about the configuration settings. `spring_cloud_dataflow_version-info_type` seems to allow only one type of checksum at a time.

**Comment by sabbyanandan** [comment]
Sorry for the back and forth here. This doesn't have to be complicated. The logic to hide extra attributes from the response could be avoided.  What if we return just SHA-1 for now and revisit the SHA-256 requirement later when we know it's available? _(we can add a separate story to track that; it also helps making progress on this story)_

**Comment by cppwfs** [comment]
Not trying to muddy the waters, but I can make it configurable for each of the artifacts. i.e. 
spring_cloud_dataflow_version-info_shell_url = where your artifact is located
spring_cloud_dataflow_version-info_shell_checksum = sha1
And in the result return something like this:
```
"shell": {
      "name": "Spring Cloud Data Flow Shell",
      "version": "1.3.0.BUILD-SNAPSHOT",
      "checksum": "bfcd1b73089b8f07b2a201036865b25a4624ab19",
      "checksum-type": "sha1",
      "url": "https:\/\/repo.spring.io\/libs-snapshot\/org\/springframework\/cloud\/spring-cloud-dataflow-shell\/1.3.0.BUILD-SNAPSHOT\/spring-cloud-dataflow-shell-1.3.0.BUILD-SNAPSHOT.jar"
 }
```

**Comment by glyn** [comment]
@sabbyanandan That suits me fine, providing we can agree the JSON format that will be produced when SHA-256 _is_ available. Then I can write the plugin now and it will _just work_ when SHA-256 is available.

**Comment by glyn** [comment]
@cppwfs and I discussed the JSON format. He is going to explore how to implement returning this format in the short term:
```
"shell": {
      "name": "Spring Cloud Data Flow Shell",
      "version": "1.3.0.BUILD-SNAPSHOT",
      "checksum-sha1": "bfcd1b73089b8f07b2a201036865b25a4624ab19",
      "url": "https:\/\/repo.spring.io\/libs-snapshot\/org\/springframework\/cloud\/spring-cloud-dataflow-shell\/1.3.0.BUILD-SNAPSHOT\/spring-cloud-dataflow-shell-1.3.0.BUILD-SNAPSHOT.jar"
 }
```
and the following (backwards compatible) format when SHA-256 becomes available:
```
"shell": {
      "name": "Spring Cloud Data Flow Shell",
      "version": "1.3.0.BUILD-SNAPSHOT",
      "checksum-sha1": "bfcd1b73089b8f07b2a201036865b25a4624ab19",
      "checksum-sha256": "aabbccddeeffaabbccddeeffaabbccddeeffaabbccddeeffaabbccddeeff",
      "url": "https:\/\/repo.spring.io\/libs-snapshot\/org\/springframework\/cloud\/spring-cloud-dataflow-shell\/1.3.0.BUILD-SNAPSHOT\/spring-cloud-dataflow-shell-1.3.0.BUILD-SNAPSHOT.jar"
 }
```
Note that the latter includes both types of checksum so that clients who depend on SHA-1 will not break when the SHA-256 support is shipped.

**Comment by jvalkeal** [comment]
Re-opened due to #1829
